### PR TITLE
contracts-bedrock: tightly pack creation timestamp

### PR DESCRIPTION
**Pull request by tynes**
**Description**

Pack the creation timestamp in the storage slot along with the address of the dispute games so that the
information can be accessed easily by services that query the contract.

A service looking for recent dispute games can
binary search or linear search backwards using the `gameAtIndex` function. It is guaranteed that the
timestamp will always increase or be equal as the
index increases.

Closes CLI-4114

### CODE REVIEW

File: packages/contracts-bedrock/contracts/dispute/DisputeGameFactory.sol [hunk 1]
```diff
@@ -164,4 +184,20 @@ contract DisputeGameFactory is OwnableUpgradeable, IDisputeGameFactory, IVersion
         gameImpls[gameType] = impl;
         emit ImplementationSet(address(impl), gameType);
     }
+
+    /**
+     * @dev Packs an address and a uint256 into a single bytes32 slot.
+     */
+    function _packSlot(address _addr, uint256 _num) internal pure returns (bytes32 slot_) {
+        slot_ = bytes32(_num << 160) | bytes32(uint256(uint160(address(_addr))));
+    }
```
Comment (refcell) [inline packages/contracts-bedrock/contracts/dispute/DisputeGameFactory.sol:193]: Can we fuzz test this?

**Review by refcell — CHANGES_REQUESTED** [review]
Would like to see fuzz test on the slot packing + looks like some lines exceed the ci lint limit

File: packages/contracts-bedrock/contracts/dispute/DisputeGameFactory.sol [hunk 2]
```diff
@@ -29,18 +32,18 @@ contract DisputeGameFactory is OwnableUpgradeable, IDisputeGameFactory, IVersion
...
 
     /**
-     * @notice Mapping of a hash of `gameType . rootClaim . extraData` to
+     * @notice Mapping of a hash of `gameType || rootClaim || extraData` to
      *         the deployed `IDisputeGame` clone.
-     * @dev Note: `.` denotes concatenation.
+     * @dev Note: `||` denotes concatenation.
      */
-    mapping(Hash => IDisputeGame) internal disputeGames;
+    mapping(Hash => bytes32) internal _disputeGames;
```
[the remark this answers is not in the document]
Comment (maurelian) [reply packages/contracts-bedrock/contracts/dispute/DisputeGameFactory.sol:39]: I think this is more consistent with the usage of custom types elsewhere (which I appreciate).

```suggestion
    type GameId is bytes32;

    mapping(Hash => GameId) internal _disputeGames;
```

File: packages/contracts-bedrock/contracts/dispute/DisputeGameFactory.sol (hunk 1 above)
Comment (maurelian) [reply packages/contracts-bedrock/contracts/dispute/DisputeGameFactory.sol:193]: +1. 
You could basically copy, paste and lightly modify the [Aliasing invariant tests](https://github.com/ethereum-optimism/optimism/blob/7015d66c9f5302a92b0ead60dc41fd05d350ade6/packages/contracts-bedrock/contracts/test/invariants/AddressAliasHelper.t.sol#L1).

File: packages/contracts-bedrock/contracts/dispute/DisputeGameFactory.sol [hunk 3]
```diff
@@ -61,27 +64,42 @@ contract DisputeGameFactory is OwnableUpgradeable, IDisputeGameFactory, IVersion
...
     function games(
-        GameType gameType,
-        Claim rootClaim,
-        bytes calldata extraData
-    ) external view returns (IDisputeGame _proxy) {
-        return disputeGames[getGameUUID(gameType, rootClaim, extraData)];
+        GameType _gameType,
+        Claim _rootClaim,
+        bytes calldata _extraData
+    ) external view returns (IDisputeGame proxy_, uint256 timestamp_) {
```
Comment (maurelian) [inline packages/contracts-bedrock/contracts/dispute/DisputeGameFactory.sol:87]: My primary concern with named returns is that we're breaking convention with the rest of the codebase.
I also don't think they are always more readable. 

IMO they are most justifiable when yul is being used, and applying them in only those cases provides an opportunity to introduce them without breaking existing conventions.

File: packages/contracts-bedrock/contracts/dispute/DisputeGameFactory.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (tynes) [reply packages/contracts-bedrock/contracts/dispute/DisputeGameFactory.sol:39]: Good callout

File: packages/contracts-bedrock/contracts/dispute/DisputeGameFactory.sol [hunk 4]
```diff
@@ -29,18 +32,18 @@ contract DisputeGameFactory is OwnableUpgradeable, IDisputeGameFactory, IVersion
     mapping(GameType => IDisputeGame) public gameImpls;
 
     /**
-     * @notice Mapping of a hash of `gameType . rootClaim . extraData` to
+     * @notice Mapping of a hash of `gameType || rootClaim || extraData` to
      *         the deployed `IDisputeGame` clone.
```
[the remark this answers is not in the document]
Comment (clabby) [reply packages/contracts-bedrock/contracts/dispute/DisputeGameFactory.sol:36]: ^ Maybe `@inheritdoc` for these as to not duplicate, or get rid of the `IDisputeGameFactory` interface alltogether? (There won't be multiple implementations of the factory.)

File: packages/contracts-bedrock/contracts/dispute/DisputeGameFactory.sol (hunk 3 above)
Comment (maurelian) [reply packages/contracts-bedrock/contracts/dispute/DisputeGameFactory.sol:87]: > Given that these contracts have 0 dependencies on existing contracts within contracts-bedrock and vice-versa, would you be open to allowing a few of these updated style suggestions to go through?

With that understanding, yes.

File: packages/contracts-bedrock/contracts/dispute/DisputeGameFactory.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (tynes) [reply packages/contracts-bedrock/contracts/dispute/DisputeGameFactory.sol:36]: I think we can follow up in a PR that gets rid of the interface here

File: packages/contracts-bedrock/contracts/test/DisputeGameFactory.t.sol
[the remark this answers is not in the document]
Comment (tynes) [reply packages/contracts-bedrock/contracts/test/DisputeGameFactory.t.sol:1]: Ok sorry about that, will fix

**Review by refcell — APPROVED** [review]

**Review by clabby — APPROVED** [review]

**Pull request merged**